### ISSUE TITLE
Generic should expand in union

### ISSUE THREAD

**Issue by kiliancs** [microsoft/TypeScript#43585]
# Bug Report

### 🔎 Search Terms

generic expansion, union expansion

### 🕗 Version & Regression Information

- This changed between versions 4.1.5 and 4.2.3 (and the new behavior remains in 4.3.0-dev.20210407)

### ⏯ Playground Link

* [Playground link with an example of the regression](https://www.typescriptlang.org/play?jsx=0&ts=4.2.3#code/C4TwDgpgBAsghiARhAPAFQHxQLwCgoFQA+UAdgK4A2l+hJauuAxgPakDOwUAluwEoR2LSgDcIAExxR0GABQBbBMgBcsJakwBKVYqTReUNDgy0Cu5FACE2bGSqUA3I1YcuisFPQAaKAFU5plDmEKrwejJege4AYqSqsiJwlOQhhprGfpHaauH+GbwCQqISCurpAPxBcGCxpXrpOupOuAD0LX7s3KQA5lBwpH0AToMIPAPAABYGEAAecPJglNDITHDk7PpucADWglCT0ABmcNzJg9AsiCLcLOtQAO4TEAPsYOdw4l293MCRbVCIchcA5QN6XJaKYDcJgAiATODXFiDHjsMgsLivCBMbiHaH7FhDEYgdgAOlw4ixlDg5ygLk4UEOLBYoXUKDCyBQnEGXwA2gBdDAYJzuWSMlg+LjYLA8kmy4B8zQOKD-LgGLm8vljKAAFhJAEYfOzUOqevysLJ+pIIMMkVAAEzagAcjvSXR1JLtZJFYp8smAqhN3X56Sl+0VyvaqtRgf5WsQ6Im7r1fVIkl1doeP0TIP6pHRcChbBJQA)
* [Playground link with a workaround](https://www.typescriptlang.org/play?jsx=0&ts=4.2.3#code/C4TwDgpgBAsghiARhAPAFQHxQLwCgoFQA+UAdgK4A2l+hJauuAxgPakDOwUAluwEoR2LSgDcIAExxR0GABQBbBMgBcsJakwBKVYqTReUNDgy0Cu5FACE2bGSqUA3IwD0zwyyhxgwCPLBdgDwAnQWExKGAAC312dnJ9UlxXKEjvMHZlVwB3HIA6UEh2JiDuf0o4UgBzXJYgyucwcpAAfgArdgAPbAAGADJgdmwAFlyAJlyAZgBiVnEIZwBhIbQsgBFKsEqAIQBBdkrI7h2+SJ2ABR2AMQBFAAkO64AZLIXKlhudgGoAVR3xSoA0kMdqNKJ9IgApODkcg7DqVM5wADWqwA8llfpRyFkAKIsPijR4AZUqqyYAEkdjiOh0+N0AOI7LbXLaILYwbZMdhQpFZLYAgBqAE0AI5oATfABy6xA3QW5AArJcdgscaNEPSiSLfhNyQBGIVMci8S5na47OAAq7fBWNVbyHEAoJZCCtCEQSoQYGirXsbgQrKUVHXSKRMBBekAMzAtzg0Nun3piBe1xF5KJ8qCYHOHS2THpCw6YAAGlnUeQwOWdqtuo9iwB2dgTAHmhWULZZGPdHZoFgwckXC5MoX0nE7HYQJhnCGVSjS8loW6rJH1s5gchMasAdVnCzQ3a28iYkoAXjOAJx7bgLR5MR6o4Dm1HLMdndhC1HiLJDSijc8TGAFj4RABQWckmEiTcFWuSUADZiyhAEhTJGcdiONBUVWYtLm4VZbnaDl2EebgRAWdlEG4WNbnrS5IlWHEhUqf4dlRShv24DpKHWBUQEqR4kSGb4cVRR4tiFFhi2+AFVgWEBmVIHF6WLLIQW2cRVh2IUIRPcgtyJGBKEqT5HlaIVHh2AAOJF5BAIYtgsk9rlRaSAFpHkqelHnkCyREoVorSuSEqSFSNURPb5URFCARBAIiIU+MByXJGAYCRAVxxxE84EqHYmFaEQiWLPg9QhR5gAALQhS4hTACz2HkHYRUiSUJmLBVYKJMFugFEARDgEVgD4SpIwVLd2GuWDyXkesgj1SMBXJJFKD1cQWAgM5ujQAdYJjPgmAWSI2Qhc0kjcF0oGKCAvGgCooHIUgsiCOAwEgIIInACAABpPFISRyHYfQuG4UgoAAA0UMBQagZBKBYLJclwApoG+B6nrAeA9HQKAIA6HxfvYNRMfupFSDh0gMCwWwjBxvHxAJjHkBQUweFISMIDe75cBMQgoGaKBqSYLE5hQb5vooagsGSLJIhAC6KgAci4c7Wn+rhVrB75QeabXmdUUhovZpxmDYTgoAhqR0G+745GZ8wIFUBmNAwT7beey5SFUWQ+qxe2oBRx7nsdmRNGMP2Xe0QnGet0PeAEIRRAkBR1BDvmIfdpO9E8An-bRoOtCgHR1CNuZBbgEILpNrhIxYFgHfUFAg84EoqgAbQAXQppwIdkauWG+rhsCwFvchH4A280BwoGSIGCab4HKnb76hDu1HnrAeeoCyWokQkFw3C2cguFJrIIkOAm2GgRAghYHflFwSN7qYYBuDYbGOjgPxKFQAU5BEOvMZ-iHAA3szL+XBvbxH-ozH+TgeasA4FwEIcBJC2BQJcN+tMCY7xACwSMUAf4924BASg4hVCXAjkHAULdLgd2MMzHm3cIFfSgCIUOIgW6RiISQ8eU83BoA+lAeWlx5Zy1IKTLgFh-oSAiB4YGcwOjvUgIInOgd64-3lrkWQowFQTFgpoXAABfIAA), that causes other problems, but still probably relevant.

### 💻 Code

```ts
type Maybe<T> =
    | null
    | T

const isResolved = <T>(maybe: Maybe<T>): maybe is T =>
    maybe !== null;

const map = <T, U>(
    maybe: Maybe<T>,
    mapFn: (value: T) => U,
): Maybe<U> => isResolved(maybe) ? mapFn(maybe) : maybe;

// Using an array in this example because it makes the failure obvious when spreading it,
// but the problematic behavior is not specific to arrays.
declare const foo: Maybe<Maybe<string[]>>;
map(foo, t => [...t]); // t is string[] in 4.1, Maybe<string[]> (and error 2488) in 4.2.
map(foo, (t: string[]) => t); // t is string[] in both 4.1 and 4.2 with the annotation.
```

### 🙁 Actual behavior

`t` in `map(foo, t => [...t]);` is a `Maybe<string[]>`.

It seems like when `T` in `Maybe` is another `Maybe`, it's opaque (not expanded).

### 🙂 Expected behavior
`t` in `map(foo, t => [...t]);` is a `string[]`.

It seems like when `T` in `Maybe` is another `Maybe`, it's expanded. For example `Maybe<Maybe<string[]>`> is expanded to `null | (null | string [])`, then collapsed to `null | string[]`, which is the same as `Maybe<string[]>`.

It seems like this is the desired behavior, because at runtime, `t` will indeed only ever be `string[]`.

**Comment by weswigham** [comment]
If you check the quickinfo on `foo` the reason why becomes readily apparent - since we preserve union origins now, `null | string[]` at `foo` has an origin of `Maybe<Maybe<string[]>>`. In inference we match this against `Maybe<T>`, thus unwrapping it to `Maybe<string[]>`. Before we preserved that origin, we'd record `foo` as just `Maybe<string[]>`, (since the second `Maybe` wrap is a noop and results in an identical type) and match against `Maybe<T>` and get `string[]`.

@ahejlsberg this seems like another instance where our new union aliases happen to be producing worse inference results. I know you've looked into this a bit already - did you have an opinion here?

**Comment by ahejlsberg** [comment]
It's a tricky situation. The issue is that when types are "lossy", i.e. when two or more input types produce the same output type (as is the case for `Maybe<T>` with `string[]` or `string[] | null` as an input type), then inference results are sensitive to whether we preserved a type alias or not. In the vast number of cases it is better to preserve the alias, but here we have types that explicitly depends on the lossy behavior. My first suggestion is to change the declaration of `Maybe<T>` to

```ts
type Maybe<T> = [T | null][0];
```

which will cause us to *not* preserve the alias. An alternative is to change the declaration of `map` to

```ts
const map = <T, U>(
    maybe: T | null,
    mapFn: (value: T) => U,
): Maybe<U> => isResolved(maybe) ? mapFn(maybe) : maybe;
```

In other words, force inference to be *structural* by manually writing out the type of the `maybe` parameter.

**Comment by ahejlsberg** [comment]
I will readily admit that both of those suggestions are *subtle*, but I'm not sure I can come up with anything better.

**Comment by kiliancs** [comment]
Thanks @ahejlsberg . We went with your first suggestion.

I think it's not too bad as a fix. I agree it's subtle, and it's difficult to know the (changing) heuristics of TypeScript to opt into one behavior or another. It might be a good opportunity for (advanced) documentation.
